prometheus-node-exporter-lua: hostapd_stations metric and label makeover

mweinelt opened the pull request:
Maintainer: @blocktrron 
Compile tested: ipq40xx-generic, gl-inet b1300
Run tested: ipq40xx-generic, gl-inet b1300

Description: Correlating data is only possible when we have more metadata, this adds a bunch of labels for each client, that will increase the depth of dashboards to come.

Before:
```
hostapd_station_rx_bytes{station="24:62:ab:41:9e:28",ifname="radio0_vlan120"} 13768
hostapd_station_tx_bytes{station="24:62:ab:41:9e:28",ifname="radio0_vlan120"} 8317
hostapd_station_tx_packets{station="24:62:ab:41:9e:28",ifname="radio0_vlan120"} 94
hostapd_station_rx_packets{station="24:62:ab:41:9e:28",ifname="radio0_vlan120"} 118
hostapd_station_connected_time{station="24:62:ab:41:9e:28",ifname="radio0_vlan120"} 129
hostapd_station_signal{station="24:62:ab:41:9e:28",ifname="radio0_vlan120"} -27
hostapd_station_vht{station="24:62:ab:41:9e:28",ifname="radio0_vlan120"} 1
hostapd_station_ht{station="24:62:ab:41:9e:28",ifname="radio0_vlan120"} 1
hostapd_station_wmm{station="24:62:ab:41:9e:28",ifname="radio0_vlan120"} 1
hostapd_station_mfp{station="24:62:ab:41:9e:28",ifname="radio0_vlan120"} 1
hostapd_station_inactive_msec{station="24:62:ab:41:9e:28",ifname="radio0_vlan120"} 3820
hostapd_station_wpa{station="24:62:ab:41:9e:28",ifname="radio0_vlan120"} 2
```

After:
```
hostapd_station_flag_assoc{ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",channel="9",bssid="e4:95:6e:41:f7:a0",station="24:62:ab:41:9e:28",frequency="2452"} 1
hostapd_station_flag_short_preamble{ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",channel="9",bssid="e4:95:6e:41:f7:a0",station="24:62:ab:41:9e:28",frequency="2452"} 0
hostapd_station_flag_ht{ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",channel="9",bssid="e4:95:6e:41:f7:a0",station="24:62:ab:41:9e:28",frequency="2452"} 1
hostapd_station_flag_vht{ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",channel="9",bssid="e4:95:6e:41:f7:a0",station="24:62:ab:41:9e:28",frequency="2452"} 0
hostapd_station_flag_he{ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",channel="9",bssid="e4:95:6e:41:f7:a0",station="24:62:ab:41:9e:28",frequency="2452"} 0
hostapd_station_flag_wmm{ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",channel="9",bssid="e4:95:6e:41:f7:a0",station="24:62:ab:41:9e:28",frequency="2452"} 1
hostapd_station_flag_mfp{ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",channel="9",bssid="e4:95:6e:41:f7:a0",station="24:62:ab:41:9e:28",frequency="2452"} 0
hostapd_station_receive_packets_total{ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",channel="9",bssid="e4:95:6e:41:f7:a0",station="24:62:ab:41:9e:28",frequency="2452"} 2314
hostapd_station_receive_bytes_total{ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",channel="9",bssid="e4:95:6e:41:f7:a0",station="24:62:ab:41:9e:28",frequency="2452"} 206025
hostapd_station_transmit_packets_total{ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",channel="9",bssid="e4:95:6e:41:f7:a0",station="24:62:ab:41:9e:28",frequency="2452"} 2644
hostapd_station_transmit_bytes_total{ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",channel="9",bssid="e4:95:6e:41:f7:a0",station="24:62:ab:41:9e:28",frequency="2452"} 214329
hostapd_station_inactive_seconds{ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",channel="9",bssid="e4:95:6e:41:f7:a0",station="24:62:ab:41:9e:28",frequency="2452"} 4.14
hostapd_station_signal_dbm{ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",channel="9",bssid="e4:95:6e:41:f7:a0",station="24:62:ab:41:9e:28",frequency="2452"} -51
hostapd_station_connected_seconds_total{ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",channel="9",bssid="e4:95:6e:41:f7:a0",station="24:62:ab:41:9e:28",frequency="2452"} 5067
hostapd_station_vht_capb_su_beamformee{ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",channel="9",bssid="e4:95:6e:41:f7:a0",station="24:62:ab:41:9e:28",frequency="2452"} 0
hostapd_station_vht_capb_mu_beamformee{ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",channel="9",bssid="e4:95:6e:41:f7:a0",station="24:62:ab:41:9e:28",frequency="2452"} 0
```

champtar commented on utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/hostapd_stations.lua:
```diff
@@ -1,14 +1,42 @@
+        mode = cfg['mode'],
+        encryption = cfg['encryption'],
+        frequency = hostapd_clients['freq'],
+        channel = freq_to_chan(hostapd_clients['freq'])
```
as we compute it from the frequency this is redundant data, can't you compute it while displaying ?

mweinelt replied:
It is cheap to store both, since they always map to the same value. This results in a low cardinality within Prometheus.

Doing this conversion during Prometheus queries is somewhat harder and I don't have an idea on how to do it right now at all. If at all possible it would make queries much more complex.

mweinelt commented:
> hostapd_station_vht{ifname="radio0_vlan120",ssid="home.lossy.network",encryption="psk2",channel="9",mode="ap",station="24:62:ab:41:9e:28",frequency="2452"} 1

A bit surprised how vht could be 1 on an ESP8266 on a 2.4 GHz channel. :thinking:

mweinelt commented:
```
Thu Jun 24 03:05:03 2021 daemon.info prometheus-node-exporter-lua[40: ...r/lib/lua/prometheus-collectors/hostapd_stations.lua:128: table index is nil
```

That is the following line:

```
        current_station_values[name] = value
```

mweinelt commented:
The parsing of flags was broken. It would only set the metric if the flag was present, never when it wasn't. It also checked for character classes instead of literal values.

I rewrote all of it by parsing all flags into a set and then rendering out 0 or 1 depending on their existance in that set.

mweinelt commented on utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/hostapd_stations.lua:
```diff
@@ -1,14 +1,42 @@
+      local labels = {
+        ifname = cfg['ifname'],
+        ssid = cfg['ssid'],
+        bssid = cfg['macaddr'],
```
That is only available, when it is actually set in the config. So I feel we should probably look at `hostapd_cli -i $ifname status` and parse more information from there.

mweinelt commented:
Final result

```
hostapd_station_receive_bytes_total{channel="9",ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",bssid="e4:95:6e:41:f7:a0",station="48:3f:da:7e:f4:85",frequency="2452"} 41775231
hostapd_station_transmit_bytes_total{channel="9",ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",bssid="e4:95:6e:41:f7:a0",station="48:3f:da:7e:f4:85",frequency="2452"} 20063030
hostapd_station_transmit_packets_total{channel="9",ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",bssid="e4:95:6e:41:f7:a0",station="48:3f:da:7e:f4:85",frequency="2452"} 181680
hostapd_station_receive_packets_total{channel="9",ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",bssid="e4:95:6e:41:f7:a0",station="48:3f:da:7e:f4:85",frequency="2452"} 97957
hostapd_station_connected_seconds_total{channel="9",ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",bssid="e4:95:6e:41:f7:a0",station="48:3f:da:7e:f4:85",frequency="2452"} 148661
hostapd_station_signal_dbm{channel="9",ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",bssid="e4:95:6e:41:f7:a0",station="48:3f:da:7e:f4:85",frequency="2452"} -69
hostapd_station_flags{channel="9",ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",flag="VHT",bssid="e4:95:6e:41:f7:a0",station="48:3f:da:7e:f4:85",frequency="2452"} 0
hostapd_station_flags{channel="9",ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",flag="HT",bssid="e4:95:6e:41:f7:a0",station="48:3f:da:7e:f4:85",frequency="2452"} 0
hostapd_station_flags{channel="9",ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",flag="WMM",bssid="e4:95:6e:41:f7:a0",station="48:3f:da:7e:f4:85",frequency="2452"} 0
hostapd_station_flags{channel="9",ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",flag="MFP",bssid="e4:95:6e:41:f7:a0",station="48:3f:da:7e:f4:85",frequency="2452"} 0
hostapd_station_inactive_seconds{channel="9",ssid="home.lossy.network",vif="radio0_vlan120",encryption="psk2",bssid="e4:95:6e:41:f7:a0",station="48:3f:da:7e:f4:85",frequency="2452"} 0.11
```

mweinelt commented on utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/hostapd_stations.lua:
```diff
@@ -1,90 +1,117 @@
-      table.insert(interfaces, intf['ifname'])
+      local cfg = intf['config']
+
+      -- Migrate this to ubus interface once it exposes all interesting labels
```
https://patchwork.ozlabs.org/project/openwrt/patch/20210625220108.151779-1-hexa@darmstadt.ccc.de/

Wrote some code to that end.

champtar replied:
This just got merged https://git.openwrt.org/?p=openwrt/openwrt.git;a=commitdiff;h=398df6275674b637bc950e4f946f84c2bc5e09ff

mweinelt commented:
And another issue is that some metrics don't appear when hostapd omits kv-pairs, like `vht_capab_info` on a 2.4 GHz 802.11n radio.

Gonna have to rework more metric handling. Stay tuned.

mweinelt commented:
And finally free of race conditione, off-by-ones and optional metrics, except for `sae_group`, but I think that's fine!

mweinelt commented:
![image](https://user-images.githubusercontent.com/131599/123863428-e1327900-d929-11eb-994d-bb6feaa46364.png)

Quite happy with the station table result.

```json
{
  "id": 3,
  "gridPos": {
    "h": 16,
    "w": 24,
    "x": 0,
    "y": 60
  },
  "type": "table",
  "title": "Stations",
  "transformations": [
    {
      "id": "seriesToColumns",
      "options": {
        "byField": "station"
      }
    },
    {
      "id": "organize",
      "options": {
        "excludeByName": {
          "Time": true,
          "Time 1": true,
          "Time 2": true,
          "Time 3": true,
          "Time 4": true,
          "Time 5": true,
          "__name__": true,
          "__name__ 1": true,
          "__name__ 2": true,
          "__name__ 3": true,
          "__name__ 4": true,
          "__name__ 5": true,
          "bssid": true,
          "bssid 1": true,
          "bssid 2": true,
          "bssid 3": true,
          "bssid 4": true,
          "bssid 5": true,
          "channel 1": false,
          "channel 2": true,
          "channel 3": true,
          "channel 4": true,
          "channel 5": true,
          "encryption 1": false,
          "encryption 2": true,
          "encryption 3": true,
          "encryption 4": true,
          "encryption 5": true,
          "flag": true,
          "frequency": true,
          "frequency 1": true,
          "frequency 2": true,
          "frequency 3": true,
          "frequency 4": true,
          "frequency 5": true,
          "instance 1": true,
          "instance 2": true,
          "instance 3": true,
          "instance 4": true,
          "instance 5": true,
          "job": true,
          "job 1": true,
          "job 2": true,
          "job 3": true,
          "job 4": true,
          "job 5": true,
          "ssid 1": false,
          "ssid 2": true,
          "ssid 3": true,
          "ssid 4": true,
          "ssid 5": true,
          "vif": true,
          "vif 1": true,
          "vif 2": true,
          "vif 3": true,
          "vif 4": true,
          "vif 5": true,
          "vif 7": true,
          "ssid 7": true,
          "job 7": true,
          "instance 7": true,
          "encryption 7": true,
          "frequency 7": true,
          "channel 7": true,
          "bssid 7": true,
          "__name__ 7": true,
          "Time 7": true,
          "vif 6": true,
          "ssid 6": true,
          "job 6": true,
          "instance 6": true,
          "frequency 6": true,
          "encryption 6": true,
          "channel 6": true,
          "__name__ 6": true,
          "Time 6": true,
          "bssid 6": true
        },
        "indexByName": {
          "ssid 1": 0,
          "station": 1,
          "Value #F": 2,
          "Value #G": 3,
          "Value #E": 4,
          "encryption 1": 5,
          "channel 1": 6,
          "Value #A": 7,
          "Value #B": 8,
          "Value #C": 9,
          "Value #D": 10,
          "Time 1": 11,
          "__name__ 1": 12,
          "bssid 1": 13,
          "frequency 1": 14,
          "instance 1": 15,
          "job 1": 16,
          "vif 1": 17,
          "Time 2": 18,
          "__name__ 2": 19,
          "bssid 2": 20,
          "channel 2": 21,
          "encryption 2": 22,
          "frequency 2": 23,
          "instance 2": 24,
          "job 2": 25,
          "ssid 2": 26,
          "vif 2": 27,
          "Time 3": 28,
          "__name__ 3": 29,
          "bssid 3": 30,
          "channel 3": 31,
          "encryption 3": 32,
          "frequency 3": 33,
          "instance 3": 34,
          "job 3": 35,
          "ssid 3": 36,
          "vif 3": 37,
          "Time 4": 38,
          "__name__ 4": 39,
          "bssid 4": 40,
          "channel 4": 41,
          "encryption 4": 42,
          "frequency 4": 43,
          "instance 4": 44,
          "job 4": 45,
          "ssid 4": 46,
          "vif 4": 47,
          "Time 5": 48,
          "__name__ 5": 49,
          "bssid 5": 50,
          "channel 5": 51,
          "encryption 5": 52,
          "frequency 5": 53,
          "instance 5": 54,
          "job 5": 55,
          "ssid 5": 56,
          "vif 5": 57,
          "Time 6": 58,
          "__name__ 6": 59,
          "bssid 6": 60,
          "channel 6": 61,
          "encryption 6": 62,
          "frequency 6": 63,
          "instance 6": 64,
          "job 6": 65,
          "ssid 6": 66,
          "vif 6": 67,
          "Time 7": 68,
          "__name__ 7": 69,
          "bssid 7": 70,
          "channel 7": 71,
          "encryption 7": 72,
          "frequency 7": 73,
          "instance 7": 74,
          "job 7": 75,
          "ssid 7": 76,
          "vif 7": 77
        },
        "renameByName": {
          "Value #A": "Signal",
          "Value #B": "TX",
          "Value #C": "RX",
          "Value #D": "Connected",
          "Value #E": "MFP",
          "bssid": "",
          "channel": "Channel",
          "channel 1": "Channel",
          "encryption": "Encryption",
          "encryption 1": "Encryption",
          "frequency": "",
          "frequency / 1000": "",
          "frequency_hz": "Frequency",
          "instance": "AP",
          "job 1": "",
          "ssid": "SSID",
          "ssid 1": "SSID",
          "station": "MAC",
          "vif": "",
          "Value #G": "Association",
          "Value #F": "Authentication",
          "vif 6": ""
        }
      }
    }
  ],
  "pluginVersion": "8.0.3",
  "fieldConfig": {
    "defaults": {
      "custom": {
        "align": "auto",
        "displayMode": "auto"
      },
      "thresholds": {
        "mode": "absolute",
        "steps": [
          {
            "color": "red",
            "value": null
          },
          {
            "color": "orange",
            "value": -75
          },
          {
            "color": "green",
            "value": -65
          }
        ]
      },
      "mappings": [],
      "color": {
        "mode": "thresholds"
      },
      "unit": "none"
    },
    "overrides": [
      {
        "matcher": {
          "id": "byName",
          "options": "Connected"
        },
        "properties": [
          {
            "id": "unit",
            "value": "dtdurations"
          },
          {
            "id": "custom.width",
            "value": 140
          }
        ]
      },
      {
        "matcher": {
          "id": "byName",
          "options": "TX"
        },
        "properties": [
          {
            "id": "unit",
            "value": "decbytes"
          },
          {
            "id": "custom.displayMode",
            "value": "basic"
          },
          {
            "id": "color",
            "value": {
              "mode": "continuous-GrYlRd"
            }
          }
        ]
      },
      {
        "matcher": {
          "id": "byName",
          "options": "RX"
        },
        "properties": [
          {
            "id": "unit",
            "value": "decbytes"
          },
          {
            "id": "custom.displayMode",
            "value": "basic"
          },
          {
            "id": "color",
            "value": {
              "mode": "continuous-GrYlRd"
            }
          }
        ]
      },
      {
        "matcher": {
          "id": "byName",
          "options": "Signal"
        },
        "properties": [
          {
            "id": "unit",
            "value": "dBm"
          },
          {
            "id": "custom.displayMode",
            "value": "color-background"
          },
          {
            "id": "thresholds",
            "value": {
              "mode": "absolute",
              "steps": [
                {
                  "color": "red",
                  "value": null
                },
                {
                  "color": "#EAB839",
                  "value": -75
                },
                {
                  "color": "green",
                  "value": -60
                }
              ]
            }
          },
          {
            "id": "custom.width",
            "value": 112
          }
        ]
      },
      {
        "matcher": {
          "id": "byName",
          "options": "Channel"
        },
        "properties": [
          {
            "id": "custom.align",
            "value": "right"
          },
          {
            "id": "custom.width",
            "value": 105
          }
        ]
      },
      {
        "matcher": {
          "id": "byName",
          "options": "Encryption"
        },
        "properties": [
          {
            "id": "custom.align",
            "value": "right"
          },
          {
            "id": "custom.width",
            "value": 102
          }
        ]
      },
      {
        "matcher": {
          "id": "byName",
          "options": "MFP"
        },
        "properties": [
          {
            "id": "unit",
            "value": "bool_yes_no"
          },
          {
            "id": "custom.displayMode",
            "value": "color-text"
          },
          {
            "id": "thresholds",
            "value": {
              "mode": "absolute",
              "steps": [
                {
                  "color": "red",
                  "value": null
                },
                {
                  "color": "green",
                  "value": 1
                }
              ]
            }
          },
          {
            "id": "mappings",
            "value": [
              {
                "type": "value",
                "options": {
                  "0": {
                    "text": "❌",
                    "index": 1
                  },
                  "1": {
                    "text": "✅ ",
                    "index": 0
                  }
                }
              }
            ]
          },
          {
            "id": "custom.align",
            "value": "center"
          },
          {
            "id": "custom.width",
            "value": 44
          }
        ]
      },
      {
        "matcher": {
          "id": "byName",
          "options": "Authentication"
        },
        "properties": [
          {
            "id": "mappings",
            "value": [
              {
                "type": "value",
                "options": {
                  "0": {
                    "text": "❌",
                    "index": 1
                  },
                  "1": {
                    "text": "✅ ",
                    "index": 0
                  }
                }
              }
            ]
          },
          {
            "id": "custom.align",
            "value": "center"
          },
          {
            "id": "custom.width",
            "value": 44
          }
        ]
      },
      {
        "matcher": {
          "id": "byName",
          "options": "Association"
        },
        "properties": [
          {
            "id": "mappings",
            "value": [
              {
                "type": "value",
                "options": {
                  "0": {
                    "text": "❌",
                    "index": 1
                  },
                  "1": {
                    "text": "✅ ",
                    "index": 0
                  }
                }
              }
            ]
          },
          {
            "id": "custom.align",
            "value": "center"
          },
          {
            "id": "custom.width",
            "value": 56
          }
        ]
      },
      {
        "matcher": {
          "id": "byName",
          "options": "Channel"
        },
        "properties": [
          {
            "id": "custom.align",
            "value": "center"
          }
        ]
      }
    ]
  },
  "options": {
    "showHeader": true,
    "frameIndex": 0,
    "sortBy": [
      {
        "displayName": "Signal",
        "desc": true
      }
    ]
  },
  "targets": [
    {
      "expr": "hostapd_station_signal_dbm{channel=~\"$channel\",ssid=~\"$ssid\",bssid=~\"$bssid\"}",
      "legendFormat": "Signal",
      "interval": "",
      "exemplar": true,
      "format": "table",
      "instant": true,
      "intervalFactor": 1,
      "refId": "A"
    },
    {
      "exemplar": true,
      "expr": "hostapd_station_transmit_bytes_total{channel=~\"$channel\",ssid=~\"$ssid\",bssid=~\"$bssid\"}",
      "format": "table",
      "hide": false,
      "instant": true,
      "interval": "",
      "legendFormat": "TX",
      "refId": "B"
    },
    {
      "exemplar": true,
      "expr": "hostapd_station_receive_bytes_total{channel=~\"$channel\",ssid=~\"$ssid\",bssid=~\"$bssid\"}",
      "format": "table",
      "hide": false,
      "instant": true,
      "interval": "",
      "legendFormat": "RX",
      "refId": "C"
    },
    {
      "exemplar": true,
      "expr": "hostapd_station_connected_seconds_total{channel=~\"$channel\",ssid=~\"$ssid\",bssid=~\"$bssid\"}",
      "format": "table",
      "hide": false,
      "instant": true,
      "interval": "",
      "legendFormat": "Connected",
      "refId": "D"
    },
    {
      "expr": "hostapd_station_flag_mfp{channel=~\"$channel\",ssid=~\"$ssid\",bssid=~\"$bssid\"}",
      "legendFormat": "MFP",
      "interval": "",
      "exemplar": true,
      "format": "table",
      "hide": false,
      "instant": true,
      "refId": "E"
    },
    {
      "expr": "hostapd_station_flag_auth{channel=~\"$channel\",ssid=~\"$ssid\",bssid=~\"$bssid\"}",
      "legendFormat": "Authentication",
      "interval": "",
      "exemplar": true,
      "refId": "F",
      "hide": false,
      "instant": true,
      "format": "table"
    },
    {
      "expr": "hostapd_station_flag_assoc{channel=~\"$channel\",ssid=~\"$ssid\",bssid=~\"$bssid\"}",
      "legendFormat": "Association",
      "interval": "",
      "exemplar": true,
      "refId": "G",
      "hide": false,
      "instant": true,
      "format": "table"
    }
  ],
  "datasource": null
}
```

champtar commented:
@mweinelt is the PR description still correct, ie before/after you are only adding labels but the metrics stay the same ?
Can you clearly document the breaking changes (or lack of) in the PR and the commit message(s) ?

Also please squash the commits (you are adding then removing hostapd_station_flags) and finally you need to bump `PKG_VERSION`

mweinelt commented:
@champtar I squashed the flag metric handling, so there is only one change to it now. The rest of the commits fulfill their own purpose and should not be squashed. I added notes, to the first and third commit, as to their changing nature.

champtar commented:
@blocktrron as this is a breaking change, are you ok with it or should we create a new exporter and mark your version as deprecated ?

blocktrron commented:
@champtar I was in contact with @mweinelt over the whole process. Breaking changes are fine with me.

champtar commented:
Thanks !
